### PR TITLE
add unitary channel

### PR DESCRIPTION
**Pull request by Roger-luo**
This will provide YaoBlocks with the some basic ability to handle noise.

### CODE REVIEW

File: test/composite/unitary_channel.jl [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using Test
+
+@testset "check apply" begin
+    r = rand_state(1)
+    channel = UnitaryChannel([X, Y, Z], [1, 0, 0])
+    @test apply!(copy(r), channel) == apply!(copy(r), X)
+end
+
+@testset "check mat" begin
+    mat(UnitaryChannel([X, Y, Z], [1, 0, 0])) == mat(X)
```
Comment (GiggleLiu) [inline test/composite/unitary_channel.jl]: why does it have a matrix? It is probablistic.

**Comment by Roger-luo** [comment]
I think printing should be tested in doc test, but somehow our doctests are all broken. I'll fix them in next release we will need this and #100 to tag next minor version

**Comment by GiggleLiu** [comment]
Great work. But still need more functions to make it work.
The example in doctest is not a typical using case. One would expect the following to work
```julia
julia> uc  = UnitaryChannel([X, Y])
nqubits: 1
unitary_channel
├─ [1.0] X gate
└─ [1.0] Y gate

julia> reg = rand_state(5)
ArrayReg{1, Complex{Float64}, Array...}
    active qubits: 5/5

julia> reg |> put(5, 2=>uc)
ERROR: unitary channel can not have a matrix
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] mat(::Type{Complex{Float64}}, ::UnitaryChannel{1,StatsBase.Weights{Float64,Float64,Array{Float64,1}}}) at /home/leo/.julia/dev/YaoBlocks/src/composite/unitary_channel.jl:92
```

It could be fixed by defining `apply!(reg, Put{N,<:UnitaryChannel})`.
Since Noise channels do not have a matrix form, their composibility is limited.
Note `Subroutine` always work for blocks with `apply!` function.

Also, I suspect it is not so useful have default values for weights (uniform)? Is it how people use unitary channels in error correction community?

**Comment by Roger-luo** [comment]
e.g you can have uniform distributed Pauli errors  `UnitaryChannel([I2, X, Y, Z])`, it's just for convenience. The simplest error model can just follow a uniform distribution I think. More complicated error model can use a set of error distribution "learnt" from experimental data: https://arxiv.org/pdf/1907.13022.pdf I'm working on something related so that's all I need for now.

**Pull request merged**